api/python/t4/bucket.py

```diff
@@ -0,0 +1,197 @@
+import pathlib
+from urllib.parse import urlparse
+
+from .data_transfer import (TargetType, copy_file, delete_object, deserialize_obj,
+                            get_bytes, get_meta, list_objects, put_bytes,
+                            serialize_obj, set_meta)
+from .util import QuiltException, fix_url, parse_s3_url
+
+class Bucket(object):
+    """
+    Implements Bucket interface for T4.
+    """
+    def __init__(self, bucket_uri):
+        """
+        Creates a Bucket object.
+
+        Args:
+            bucket_uri(str): URI of bucket to target. Must start with 's3://'
+
+        Returns:
+            a new Bucket
+        """
+        parsed = urlparse(bucket_uri)
+        bucket, _, _ = parse_s3_url(parsed)
+        self._uri = parsed.geturl().rstrip('/') + '/' # Ensure trailing '/'
+        self._bucket = bucket
+
+    def deserialize(self, key):
+        """
+        Deserializes object at key from bucket.
+
+        Args:
+            key(str): key in bucket to get
+
+        Returns:
+            deserialized object
+
+        Raises:
+            KeyError if key does not exist
+            if deserialization fails
+        """
+        data, meta = get_bytes(self._uri + key)
+        target = meta.get('target', None)
+        if not target:
+            raise QuiltException("No deserialization metadata, cannot deserialize object")
+
+        target = TargetType(target)
+        return deserialize_obj(data, target)
+
+    def __call__(self, key):
+        """
+        Shorthand for deserialize(key)
+        """
+        return self.deserialize(key)
+
+    def put(self, key, obj, meta=None):
+        """
+        Stores obj at key in bucket, optionally with user-provided metadata.
+
+        Args:
+            key(str): key in bucket to put object to
+            obj(serializable): serializable object to store at key
+            meta(dict): optional user-provided metadata to store
+        """
+        dest = self._uri + key
+        meta = meta or {}
+        data, target = serialize_obj(obj)
+        all_meta = dict(
+            target=target.value,
+            user_meta=meta
+        )
+        put_bytes(data, dest, all_meta)
+
+    def put_file(self, key, path):
+        """
+        Stores file at path to key in bucket.
+
+        Args:
+            key(str): key in bucket to store file at
+            path(str): string representing local path to file
+
+        Returns:
+            None
+
+        Raises:
+            if no file exists at path
+            if copy fails
+        """
+        dest = self._uri + key
+        copy_file(path, dest)
+
+    def put_dir(self, key, directory):
```

Comment:
Still not convinced that it's needed, given that `put_file` also handles directories... but no strong objections.